sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala

```diff
@@ -148,6 +148,17 @@ abstract class QueryStageExec extends LeafExecNode {
  */
 abstract class ExchangeQueryStageExec extends QueryStageExec {
 
+  /**
+   * This flag aims to detect if the stage materialization is started. This helps
+   * to avoid unnecessary stage materialization when the stage is canceled.
+   */
+  @transient protected val materializationStarted = new AtomicBoolean()
```

Comment:
can we set this flag in `def materialize()`?

Reply:
Yes, it is addressed.